rclcpp/src/rclcpp/utilities.cpp

```diff
@@ -42,116 +43,233 @@
-    }
+      int error_status = strerror_r(errno, error_string, error_length);
+      if (error_status != 0) {
+        throw std::runtime_error("Failed to get error string for errno: " + std::to_string(errno));
+      }
 #endif
 #else
-    strerror_s(error_string, error_length, errno);
+      strerror_s(error_string, error_length, errno);
 #endif
-    // *INDENT-OFF* (prevent uncrustify from making unnecessary indents here)
-    throw std::runtime_error(
-      std::string("Failed to set SIGINT signal handler: (" + std::to_string(errno) + ")") +
-      error_string);
-    // *INDENT-ON*
-  }
+      // *INDENT-OFF* (prevent uncrustify from making unnecessary indents here)
+      throw std::runtime_error(
+        std::string("Failed to set SIGINT signal handler: (" + std::to_string(errno) + ")") +
+        error_string);
+      // *INDENT-ON*
+    }
 
 #ifdef HAS_SIGACTION
-  return old_action;
+    return old_action;
 #else
-  return old_signal_handler;
+    return old_signal_handler;
 #endif
-}
+  }
 
-void
+  static
+  void
 #ifdef HAS_SIGACTION
-signal_handler(int signal_value, siginfo_t * siginfo, void * context)
+  signal_handler(int signal_value, siginfo_t * siginfo, void * context)
 #else
-signal_handler(int signal_value)
+  signal_handler(int signal_value)
 #endif
-{
-  RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "signal_handler(signal_value=%d)", signal_value);
+  {
+    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "signal_handler(signal_value=%d)", signal_value);
 
 #ifdef HAS_SIGACTION
-  if (old_action.sa_flags & SA_SIGINFO) {
-    if (old_action.sa_sigaction != NULL) {
-      old_action.sa_sigaction(signal_value, siginfo, context);
-    }
-  } else {
-    if (
-      old_action.sa_handler != NULL &&  // Is set
-      old_action.sa_handler != SIG_DFL &&  // Is not default
-      old_action.sa_handler != SIG_IGN)  // Is not ignored
-    {
-      old_action.sa_handler(signal_value);
+    if (old_action.sa_flags & SA_SIGINFO) {
+      if (old_action.sa_sigaction != NULL) {
+        old_action.sa_sigaction(signal_value, siginfo, context);
+      }
+    } else {
+      if (
+        old_action.sa_handler != NULL &&  // Is set
+        old_action.sa_handler != SIG_DFL &&  // Is not default
+        old_action.sa_handler != SIG_IGN)  // Is not ignored
+      {
+        old_action.sa_handler(signal_value);
+      }
     }
-  }
 #else
-  if (old_signal_handler) {
-    old_signal_handler(signal_value);
-  }
+    if (old_signal_handler) {
+      old_signal_handler(signal_value);
+    }
 #endif
+    signal_received.exchange(true);
+    events_condition_variable.notify_one();
+  }
 
-  for (auto context_ptr : rclcpp::get_contexts()) {
-    if (context_ptr->get_init_options().shutdown_on_sigint) {
-      // do not notify all, instead do that once after all are shutdown
-      context_ptr->shutdown("signal handler", false /* notify_all */);
+  void
+  deferred_signal_handler()
+  {
+    while (true) {
```

Comment:
I don't think this would work because the location of the `if (!is_installed()) {` is important, in that it comes after the loop over the context's. So I'd pass on changing this logic unless @hidmic thinks it's a good idea to do so.

Reply:
that's fine for me. I was just wondering if we could somehow change the `while(true)`, but I guess in this case it's all right.